components/brave_federated/learning_service.cc

```diff
@@ -0,0 +1,230 @@
+      std::make_unique<LearningServiceConfig>(data_resource);
+  const net::BackoffEntry::Policy post_results_policy =
+      learning_service_config_->GetPostResultsPolicy();
+  model_spec_ = std::make_unique<api::config::ModelSpec>(
+      std::move(learning_service_config_->GetModelSpec()));
+
+  post_results_policy_ =
+      std::make_unique<const net::BackoffEntry::Policy>(post_results_policy);
+  post_results_backoff_entry_ =
+      std::make_unique<net::BackoffEntry>(post_results_policy_.get());
+
+  const int init_federated_service_wait_time_in_seconds =
+      brave_federated::features::GetInitFederatedServiceWaitTimeInSeconds();
+  init_task_timer_ = std::make_unique<base::OneShotTimer>();
+  init_task_timer_->Start(
+      FROM_HERE, base::Seconds(init_federated_service_wait_time_in_seconds),
+      this, &LearningService::Init);
+}
+
+LearningService::~LearningService() {
+  if (participating_) {
+    StopParticipating();
+  }
+
+  eligibility_observation_.Reset();
+}
+
+void LearningService::Init() {
+  CHECK(url_loader_factory_);
+  CHECK(eligibility_service_);
+  CHECK(init_task_timer_);
+
+  VLOG(1) << "FL: Initializing federated learning service.";
+  eligibility_observation_.Observe(eligibility_service_);
+  if (eligibility_service_->IsEligible()) {
+    StartParticipating();
+  }
+
+  VLOG(1) << "FL: Eligibility: " << eligibility_service_->IsEligible();
+}
+
+void LearningService::OnEligibilityChanged(bool is_eligible) {
+  if (is_eligible) {
+    StartParticipating();
+    VLOG(1) << "FL: Eligibility changed, started participating.";
+  } else {
+    StopParticipating();
+    VLOG(1) << "FL: Eligibility changed, stopped participating.";
+  }
+}
+
+void LearningService::StartParticipating() {
+  if (participating_) {
+    return;
+  }
+
+  participating_ = true;
+  StartCommunicationAdapter();
+  GetTasks();
+}
+
+void LearningService::StopParticipating() {
+  if (!participating_) {
+    return;
+  }
+
+  participating_ = false;
+  communication_adapter_.reset();
+  reconnect_timer_.reset();
+}
+
+void LearningService::GetTasks() {
+  if (communication_adapter_ == nullptr) {
+    VLOG(1) << "FL: Communication adapter not initialized";
+    return;
+  }
+
+  communication_adapter_->GetTasks(base::BindOnce(
+      &LearningService::HandleTasksOrReconnect, weak_factory_.GetWeakPtr()));
+}
+
+void LearningService::HandleTasksOrReconnect(TaskList tasks,
+                                             base::TimeDelta reconnect) {
+  if (tasks.empty()) {
+    reconnect_timer_ = std::make_unique<base::RetainingOneShotTimer>();
+    reconnect_timer_->Start(FROM_HERE, reconnect, this,
+                            &LearningService::GetTasks);
+    VLOG(2) << "FL: No tasks available, reconnecting in " << reconnect;
+    return;
+  }
+
+  Task task = tasks.at(0);
+  float lr = 0.01;
+  std::map<std::string, float> config = task.GetConfig();
+  auto cursor = config.find("lr");
+  if (cursor != config.end()) {
+    lr = cursor->second;
```

Comment:
For readability it would be great to use structured bindings instead of .first / .second